relays/equivocation/src/equivocation_loop.rs

```diff
@@ -0,0 +1,327 @@
+				None
+			},
+		}
+	}
+
+	async fn synced_source_headers_at_target(
+		&mut self,
+		at: P::TargetNumber,
+	) -> Vec<HeaderFinalityInfo<P>> {
+		match self.target_client.synced_headers_finality_info(at).await {
+			Ok(synced_headers) => synced_headers,
+			Err(e) => {
+				log::error!(
+					target: "bridge",
+					"Could not get {} headers synced to {} at block {at:?}",
+					P::SOURCE_NAME,
+					P::TARGET_NAME
+				);
+
+				// Reconnect in case of a connection error.
+				self.reconnect_target_client(e).await;
+				// And move on to the next block.
+				vec![]
+			},
+		}
+	}
+
+	async fn report_equivocation(&mut self, at: P::Hash, equivocation: P::EquivocationProof) {
+		match self.reporter.submit_report(&self.source_client, at, equivocation.clone()).await {
+			Ok(_) => {},
+			Err(e) => {
+				log::error!(
+					target: "bridge",
+					"Could not submit equivocation report to {} for {equivocation:?}: {e:?}",
+					P::SOURCE_NAME,
+				);
+
+				// Reconnect source client and move on
+				self.reconnect_source_client(e).await;
+			},
+		}
+	}
+
+	async fn check_block(
+		&mut self,
+		block_num: P::TargetNumber,
+		context: &mut EquivocationReportingContext<P>,
+	) {
+		let synced_headers = self.synced_source_headers_at_target(block_num).await;
+
+		for synced_header in synced_headers {
+			self.finality_proofs_buf.fill(&mut self.finality_proofs_stream);
+
+			let equivocations = match P::EquivocationsFinder::find_equivocations(
+				&context.synced_verification_context,
+				&synced_header.finality_proof,
+				self.finality_proofs_buf.buf().as_slice(),
+			) {
+				Ok(equivocations) => equivocations,
+				Err(e) => {
+					log::error!(
+						target: "bridge",
+						"Could not search for equivocations in the finality proof \
+						for source header {:?} synced at target block {block_num:?}: {e:?}",
+						synced_header.finality_proof.target_header_hash()
+					);
+					continue
+				},
+			};
+			for equivocation in equivocations {
+				self.report_equivocation(context.synced_header_hash, equivocation).await;
+			}
+
+			self.finality_proofs_buf
+				.prune(synced_header.finality_proof.target_header_number(), None);
+			context.update(synced_header);
+		}
+	}
+
+	async fn run(&mut self, tick: Duration, exit_signal: impl Future<Output = ()>) {
+		let exit_signal = exit_signal.fuse();
+		futures::pin_mut!(exit_signal);
+
+		loop {
+			// Make sure that we are connected to the source finality proofs stream.
+			match self.finality_proofs_stream.ensure_stream(&self.source_client).await {
+				Ok(_) => {},
+				Err(e) => {
+					log::error!(
+						target: "bridge",
+						"Could not connect to the {} `FinalityProofsStream`: {e:?}",
+						P::SOURCE_NAME,
+					);
+
+					// Reconnect to the source client if needed
+					match self.reconnect_source_client(e).await {
+						true => {
+							// Connection error. Move on.
+							continue
+						},
+						false => {
+							// Irrecoverable error. End the loop.
```

Comment:
> we risk getting stuck in a retrial loop

That's true. Normally (in other relay loops) we'll eventually break of this loop when we'll read updated state and realized that something has changed. But until then we'll keep retrying. Which is fine (imo), unless we spam RPC server with failing RPC requests without some backoff.

In your examples:
- `we can't generate the key ownership prof, because we're too many blocks ahead`: then the loop should just log an error it and go further (imo);
- `let's say that the chains were updated and the relayer has old data structures, so it can't encode/decode them
etc`: that's definitely a maintenance flaw - the best we could do here is to have a backoff mechanism. Like your current loop implementation would just exit the process if you encounter any non-connection error. Is it good? E.g. if we can't generate ownership proof, then we could just keep working with next justifications.

But as I said - in the end it is up to you, how to handle issues :)